src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	K_ETW_TYPE_FLAGS_ARRAY = 0x8,
+
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_MASK = 0x3F00,
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_SHIFT = 8,
+	K_ETW_TYPE_FLAGS_ARRAY_RANK_MAX = K_ETW_TYPE_FLAGS_ARRAY_RANK_MASK >> K_ETW_TYPE_FLAGS_ARRAY_RANK_SHIFT
+} EtwTypeFlags;
+
+// This only contains the fixed-size data at the top of each struct in
+// the bulk type event.  These fields must still match exactly the initial
+// fields of the struct described in the manifest.
+typedef struct _EventStructBulkTypeFixedSizedData {
+	uint64_t type_id;
+	uint64_t module_id;
+	uint32_t type_name_id;
+	uint32_t flags;
+	uint8_t cor_element_type;
+} EventStructBulkTypeFixedSizedData;
+
+// Represents one instance of the Value struct inside a single BulkType event
+typedef struct _BultTypeValue {
+	EventStructBulkTypeFixedSizedData fixed_sized_data;
+	uint32_t c_type_parameters;
+	intptr_t rg_type_parameters[32];
+	MonoType *rg_mono_type_parameters[32];
+	char *s_name;
+} BulkTypeValue;
+
+static
+uint32_t
+ep_rt_mono_get_byte_count_in_event(BulkTypeValue *bulk_type_value)
+{
+	uint32_t s_name_len = 0;
+	if (bulk_type_value->s_name)
+		s_name_len = strlen(bulk_type_value->s_name);
+
+	return
+		sizeof(bulk_type_value->fixed_sized_data) +
+		sizeof(bulk_type_value->c_type_parameters) +
+		(s_name_len + 1) * sizeof(char) + 			// Size of name, including null terminator
+		bulk_type_value->c_type_parameters * sizeof(uint64_t);	// Type parameters
+}
+
+// ETW has a limitation of 64K for TOTAL event Size, however there is overhead associated with
+// the event headers.   It is unclear exactly how much that is, but 1K should be sufficiently
+// far away to avoid problems without sacrificing the perf of bulk processing.
+static const uint32_t cb_max_etw_event = 63 * 1024;
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array.  (Intentionally overestimate the size of the non-array parts to keep it safe.)
+static const uint32_t k_max_bytes_type_values = (cb_max_etw_event - 0x30);
+
+// Estimate of how many type value elements we can put into the struct array, while
+// staying under the ETW event size limit. Note that this is impossible to calculate
+// perfectly, since each element of the struct array has variable size.
+//
+// In addition to the byte-size limit per event, Windows always forces on us a
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+static const uint32_t k_max_count_type_values = (128 - 2) / 4;
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer[65536];
```

Comment:
I'll add it as a const along with the other globals like `cb_max_etw_event` `k_max_bytes_type_values` and `k_max_count_type_values`. I also changed the buffer to be just `uint8_t *m_p_bulk_type_event_buffer` and allocating that memory size `p_type_logger->m_p_bulk_type_event_buffer = malloc(sizeof(uint8_t) * 65536);`. I think that is more correct in terms of filling the buffer when `ep_rt_mono_fire_bulk_type_event`. 

The original form was a mistake before knowing to try to allocate memory. With the memcpy format, it would have tried to save fixed_sized_data bytes into the first 4 indices, and then jump to the 33rd and 34th index to fill in the type's string name and so forth.

I also finally discovered the bug I had been looking for the past few days where fixed_sized_data is 25 bytes (typeID - 8, moduleID - 8, typeNameID - 4, Flags - 4, CorElementType - 1), and the dotnet-pgo tool understands it as that size; however, that struct will have a sizeof `32` from padding, leading to mismatch in actual offset and expected offset of the data.

Reply:
Added the constant `k_size_of_event_buffer` (naming attributed to CoreCLR's `kSizeOfEventBuffer`)